src/reconcile/rec_visibility.c

```diff
@@ -152,29 +152,26 @@ __rec_append_orig_value(
  *     Return if we need to save the update chain
  */
 static bool
-__rec_need_save_upd(WT_SESSION_IMPL *session, WT_UPDATE *selected_upd, uint64_t max_txn,
-  wt_timestamp_t max_ts, bool has_newer_updates, uint64_t flags, WT_PAGE *page)
+__rec_need_save_upd(
+  WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_UPDATE_SELECT *upd_select, bool has_newer_updates)
 {
+    if (F_ISSET(r, WT_REC_EVICT) && has_newer_updates)
+        return (true);
+
     /*
      * Save updates for any reconciliation that doesn't involve history store (in-memory database
      * and fixed length column store), except when the maximum timestamp and txnid are globally
      * visible.
      */
-    if (LF_ISSET(WT_REC_IN_MEMORY) || page->type == WT_PAGE_COL_FIX)
-        return (!__wt_txn_visible_all(session, max_txn, max_ts));
-
-    if (!LF_ISSET(WT_REC_HS))
-        return false;
-
-    if (LF_ISSET(WT_REC_EVICT) && has_newer_updates)
-        return true;
+    if (!F_ISSET(r, WT_REC_HS) && !F_ISSET(r, WT_REC_IN_MEMORY) && r->page->type != WT_PAGE_COL_FIX)
```

Comment:
I think this is not correct. Even there is no newer updates for in memory and fix length column store, we cannot discard the update chain if there is update that is not globally visible on the update chain. The current implementation for them are very memory inefficient as they don't support history store so we have to keep all the history in memory until they are all globally visible.

Reply:
Right, for in-memory or column store we don't return here and instead fall through to the `__wt_txn_visible_all` check at line 173.  If there are newer updates, we've already returned true at line 159.  If there are no newer updates, the on-page version is the newest and we're checking its visibility.

Are you worried about some out-of-order case (i.e., from a lower isolation level), where even though the on-page version will be seen by all readers, there is some update further along the list that is not yet visible to everyone?  If so, why is that a problem?

Reply:
Oh right. I misunderstood the code. This is correct. Sorry for that.

Reply:
I confused this line with if (!F_ISSET(r, WT_REC_HS) && !F_ISSET(r, WT_REC_IN_MEMORY) && r->page->type != WT_PAGE_COL_FIX)

 if (!F_ISSET(r, WT_REC_HS) && F_ISSET(r, WT_REC_IN_MEMORY) && r->page->type = WT_PAGE_COL_FIX)